src/main/java/spoon/reflect/path/impl/CtRolePathElement.java

```diff
@@ -106,14 +54,65 @@ public CtRole getRole() {
 
 	@Override
 	public String toString() {
-		return STRING + role.toString() + getParamString();
+		return STRING + getRole().toString() + getParamString();
+	}
+
+	public CtElement getFromSet(Set set, String name) throws CtPathException {
+		for (Object o: set) {
+			if (o instanceof CtNamedElement) {
+				if (((CtNamedElement) o).getSimpleName().equals(name)) {
+					return (CtElement) o;
+				}
+			} else if (o instanceof CtReference) {
+				if (((CtReference) o).getSimpleName().equals(name)) {
+					return (CtElement) o;
+				}
+			} else {
+				throw new CtPathException();
+			}
+		}
+		throw new CtPathException();
 	}
 
 	@Override
 	public Collection<CtElement> getElements(Collection<CtElement> roots) {
-		RoleVisitor visitor = new RoleVisitor();
-		visitor.scan(roots);
-		return visitor.matchs;
+		Collection<CtElement> matchs = new LinkedList<>();
+		for (CtElement root : roots) {
+			RoleHandler roleHandler = RoleHandlerHelper.getOptionalRoleHandler(root.getClass(), getRole());
+			if (roleHandler != null) {
+				switch (roleHandler.getContainerKind()) {
+					case SINGLE:
+						matchs.add(roleHandler.getValue(root));
+						break;
+
+					case LIST:
+						if (getArguments().containsKey("index")) {
+							int index = Integer.parseInt(getArguments().get("index"));
+							matchs.add((CtElement) roleHandler.asList(root).get(index));
+						}
```

Comment:
else it should add all items of the into `matchs`. WDYT?

Reply:
and what about case when `index >= size()`?

Reply:
Yes, you are right. I only had in mind my Unique path case in mind here, but yes. I'll fix this

Reply:
Right, that also means that I should add negative test cases...

Reply:
> else it should add all items of the into `matchs`.
We should extends that behavior to sets and maps, don't you think?